arrow-buffer/src/util/bit_mask.rs

```diff
@@ -127,16 +127,16 @@ unsafe fn set_upto_64bits(
 }
 
 /// # Safety
-/// The caller must ensure all arguments are within the valid range.
+/// The caller must ensure `data` has `offset..(offset + 8)` range
 #[inline]
 unsafe fn read_bytes_to_u64(data: &[u8], offset: usize, count: usize) -> u64 {
     debug_assert!(count <= 8);
-    let mut tmp = std::mem::MaybeUninit::<u64>::new(0);
+    let mut tmp : u64 = 0;
     let src = data.as_ptr().add(offset);
     unsafe {
-        std::ptr::copy_nonoverlapping(src, tmp.as_mut_ptr() as *mut u8, count);
-        tmp.assume_init()
+        std::ptr::copy_nonoverlapping(src, &mut tmp as *mut _ as *mut u8, count);
```

Comment:
Shouldn't we restrict the number of bytes copied to 8 here? The destination is not valid for more writes.
```suggestion
        std::ptr::copy_nonoverlapping(src, &mut tmp as *mut _ as *mut u8, count.min(8));
```

Reply:
Yes, either the debug assertion should be a real assertion or we should clamp.

Reply:
_or_ the function should be documented with clearer invariants. It is an unsafe function after all, it's allowed to assume these things, but "all arguments are within range" is two separate checks: that count is not more than 8, and that offset/count is in range. The next commit adds the latter, not the former.

Reply:
Sorry, I don't know how to use github review. I meant to leave a comment and mark conversation as resolved (that's how it works with the code review tool at work), but I hit the button and it ate my comment and immediately closed the thread.

This is a good point. I don't want to change the behavior in this PR, to keep it easy to merge / a completely clear net improvement PR, so I just changed the comment.

In general, I am in favor of adding runtime checks everywhere and removing all `unsafe`, but I worry that it would defeat the point of these files, which are trying so hard to be fast and unsafe and stuff. I wouldn't know what benchmarks to check for regressions in, so I'd prefer to leave that kind of change to the code owners.